components/tools/OmeroPy/src/omero/gateway/__init__.py

```diff
@@ -141,6 +141,95 @@ def fileread_gen(fin, fsize, bufsize):
     fin.close()
 
 
+def countAnnotations(queryService, ctx, obj_type, obj_ids=[]):
+    """
+    Count the annotions linked to the given object
+
+    :param queryService:  Reference to the query service
+    :param ctx:           The security context
+    :param obj_type:      The type of the object the annotations are linked to
+    :param obj_ids:       List of object ids
+    :return:              Dictionary of annotation counts per annotation type
+    """
+
+    counts = {
+        "TagAnnotation": 0,
+        "FileAnnotation": 0,
+        "CommentAnnotation": 0,
+        "LongAnnotation": 0,
+        "MapAnnotation": 0,
+        "OtherAnnotation": 0}
+
+    if obj_type is None or not obj_ids:
+        return counts
+
+    params = omero.sys.ParametersI()
+    params.addIds(obj_ids)
+    params.add('ratingns',
+               rstring(omero.constants.metadata.NSINSIGHTRATING))
+
+    q = """
+        select sum( case when an.ns = :ratingns
+                    and an.class = LongAnnotation
+                    then 1 else 0 end),
+            sum( case when (an.ns is null or an.ns != :ratingns)
+                           and an.class = LongAnnotation
+                           then 1 else 0 end),
+           sum( case when an.class != LongAnnotation
+                    then 1 else 0 end ), an.class
+           from Annotation an where an.id in
+                (select distinct(ann.id) from %sAnnotationLink ial
+                    join ial.child as ann
+                    join ial.parent as i
+            where i.id in (:ids))
+                group by an.class
+        """ % obj_type
+
+    queryResult = queryService.projection(q, params, ctx)
+
+    for r in queryResult:
+        ur = unwrap(r)
+        if ur[3] == '/basic/num/long/':
```

Comment:
https://github.com/openmicroscopy/openmicroscopy/pull/5056 was merged, is there anything stopping you from using discriminator?